Unable to install package globally

Since the new `preinstall` npm hook was setup to configure the git hooks path, it looks like we can no longer install the CLI globally due to it relying on a `.git` directory - 

```
bscot@BSCOTT-PC ~/proj/ (feature/session-is-not-persisting-when) -> npm i -g and-cli@1.1.2

> and-cli@1.1.2 preinstall C:\Users\bscot\AppData\Roaming\npm\node_modules\and-cli
> npm run configure:git


> and-cli@1.1.2 configure:git C:\Users\bscot\AppData\Roaming\npm\node_modules\and-cli
> echo Ensuring git hooksPath is set to .githooks directory && git config core.hooksPath .githooks && chmod +x .githooks/*

Ensuring git hooksPath is set to .githooks directory
fatal: not in a git directory
npm ERR! code ELIFECYCLE
npm ERR! errno 128
npm ERR! and-cli@1.1.2 configure:git: `echo Ensuring git hooksPath is set to .githooks directory && git config core.hooksPath .githooks && chmod +x .githooks/*`
npm ERR! Exit status 128
npm ERR!
npm ERR! Failed at the and-cli@1.1.2 configure:git script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\bscot\AppData\Roaming\npm-cache\_logs\2020-08-28T12_32_33_755Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 128
npm ERR! and-cli@1.1.2 preinstall: `npm run configure:git`
npm ERR! Exit status 128
npm ERR!
npm ERR! Failed at the and-cli@1.1.2 preinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\bscot\AppData\Roaming\npm-cache\_logs\2020-08-28T12_32_34_286Z-debug.log
```

--- a/and-cli-dotnet.js
+++ b/and-cli-dotnet.js
@@ -62,7 +62,7 @@ require("./command-runner").run(async () => {
     }
 
     if (program.cli) {
-        dotnetCli.run(program.args.join(" "));
+        dotnetCli.run(program.args);
     }
 
     if (program.kill) {

--- a/modules/dotnet-cli.js
+++ b/modules/dotnet-cli.js
@@ -21,11 +21,8 @@ const shell = require("shelljs");
 
 const dotnetCli = {
     cmd(cliArgs = []) {
-        return commandStringFactory.build(
-            "dotnet",
-            dotnetPath.cliPath(),
-            ...cliArgs
-        );
+        const pathName = path.basename(dotnetPath.cliPath() || "");
+        return commandStringFactory.build("dotnet", pathName, ...cliArgs);
     },
     description() {
         return (
@@ -40,13 +37,15 @@ const dotnetCli = {
         return optionStringFactory.build("cli", "C");
     },
     run(cliArgs = []) {
-        const cliPath = dotnetPath.cliPath();
+        const cliDir = dotnetPath.cliDir();
 
         // Build dotnet project if the *Cli.dll is not found
-        if (cliPath == null) {
+        if (cliDir == null) {
             echo.warn("No Cli.dll found. Building project");
             dotnetBuild.run(true, true);
         }
+
+        dir.pushd(dotnetPath.cliDir());
 
         const { cmd, args } = this.cmd(cliArgs);
 
@@ -60,6 +59,8 @@ const dotnetCli = {
             echo.error("Command failed, see output for details.");
             shell.exit(status);
         }
+
+        dir.popd();
     },
 };
 

--- a/modules/dotnet-cli.js
+++ b/modules/dotnet-cli.js
@@ -21,10 +21,9 @@ const shell = require("shelljs");
 
 const dotnetCli = {
     cmd(cliArgs = "") {
-        const pathName = path.basename(dotnetPath.cliPath() || "");
         return commandStringFactory.build(
             "dotnet",
-            pathName,
+            dotnetPath.cliPath(),
             cliArgs.split(" ")
         );
     },
@@ -41,15 +40,13 @@ const dotnetCli = {
         return optionStringFactory.build("cli", "C");
     },
     run(cliArgs = "") {
-        const cliDir = dotnetPath.cliDir();
+        const cliPath = dotnetPath.cliPath();
 
         // Build dotnet project if the *Cli.dll is not found
-        if (cliDir == null) {
+        if (cliPath == null) {
             echo.warn("No Cli.dll found. Building project");
             dotnetBuild.run(true, true);
         }
-
-        dir.pushd(dotnetPath.cliDir());
 
         const { cmd, args } = this.cmd(cliArgs);
 
@@ -63,8 +60,6 @@ const dotnetCli = {
             echo.error("Command failed, see output for details.");
             shell.exit(status);
         }
-
-        dir.popd();
     },
 };
 

--- a/modules/dotnet-cli.js
+++ b/modules/dotnet-cli.js
@@ -20,11 +20,11 @@ const shell = require("shelljs");
 // -----------------------------------------------------------------------------------------
 
 const dotnetCli = {
-    cmd(cliArgs = "") {
+    cmd(cliArgs = []) {
         return commandStringFactory.build(
             "dotnet",
             dotnetPath.cliPath(),
-            cliArgs.split(" ")
+            ...cliArgs
         );
     },
     description() {
@@ -39,7 +39,7 @@ const dotnetCli = {
     getOptions() {
         return optionStringFactory.build("cli", "C");
     },
-    run(cliArgs = "") {
+    run(cliArgs = []) {
         const cliPath = dotnetPath.cliPath();
 
         // Build dotnet project if the *Cli.dll is not found

--- a/modules/dotnet-cli.test.js
+++ b/modules/dotnet-cli.test.js
@@ -31,7 +31,7 @@ describe("dotnetCli", () => {
     // -----------------------------------------------------------------------------------------
 
     describe("cmd", () => {
-        test("given a string array of arguments, it returns a properly command string to the Cli executable", () => {
+        test("given a string array of arguments, it returns a properly formatted command string to the Cli path with space-separated arguments", () => {
             // Arrange
             const cliPath = testUtils.randomFile();
             const cliArgs = faker.random.words(3).split(" ");
@@ -55,7 +55,7 @@ describe("dotnetCli", () => {
     describe("run", () => {
         let shellExitSpy;
         beforeEach(() => {
-            shellExitSpy = jest.spyOn(shell, "exit").mockImplementation();
+            shellExitSpy = testUtils.spyOnShellExit();
         });
 
         test("when dotnetPath.cliPath returns undefined, it calls dotnetBuild.run", () => {

--- a/modules/dotnet-cli.test.js
+++ b/modules/dotnet-cli.test.js
@@ -7,7 +7,9 @@ const dotnetBuild = require("./dotnet-build");
 const dotnetCli = require("./dotnet-cli");
 const dotnetPath = require("./dotnet-path");
 const faker = require("faker");
+const path = require("path");
 const shell = require("shelljs");
+const testUtils = require("../tests/test-utils");
 
 // #endregion Imports
 
@@ -16,6 +18,28 @@ const shell = require("shelljs");
 // -----------------------------------------------------------------------------------------
 
 describe("dotnetCli", () => {
+    // -----------------------------------------------------------------------------------------
+    // #region cmd
+    // -----------------------------------------------------------------------------------------
+
+    describe("cmd", () => {
+        test("given a space separated string of arguments, it returns a properly command string to the Cli executable", () => {
+            // Arrange
+            const cliPath = testUtils.randomFile();
+            const cliArgs = faker.random.words(3);
+            jest.spyOn(dotnetPath, "cliPath").mockImplementation(() => cliPath);
+            const expectedString = `dotnet ${cliPath} ${cliArgs}`;
+
+            // Act
+            const result = dotnetCli.cmd(cliArgs).toString();
+
+            // Assert
+            expect(result).toBe(expectedString);
+        });
+    });
+
+    // #endregion cmd
+
     // -----------------------------------------------------------------------------------------
     // #region run
     // -----------------------------------------------------------------------------------------
@@ -26,7 +50,7 @@ describe("dotnetCli", () => {
             shellExitSpy = jest.spyOn(shell, "exit").mockImplementation();
         });
 
-        test(`when dotnetPath.cliDir() path returns undefined it calls donetBuild.run`, () => {
+        test("when dotnetPath.cliDir returns undefined, it calls dotnetBuild.run", () => {
             // Arrange
             const dotnetPathSpy = jest
                 .spyOn(dotnetPath, "cliDir")
@@ -43,14 +67,10 @@ describe("dotnetCli", () => {
             expect(dotnetBuildSpy).toHaveBeenCalled();
         });
 
-        test(`when spawn.sync returns non 0 it calls shell.exit with the status`, () => {
+        test("when child_process.spawnSync returns non-zero status, it calls shell.exit with the status", () => {
             // Arrange
-            const exitCode = faker.random.number({ min: 1 });
-            const spawnSync = jest
-                .spyOn(child_process, "spawnSync")
-                .mockImplementation(() => {
-                    return { status: exitCode };
-                });
+            const exitCode = testUtils.randomNumber(1);
+            const spawnSync = testUtils.spyOnSpawnSync(exitCode);
 
             // Act
             dotnetCli.run();

--- a/modules/dotnet-cli.test.js
+++ b/modules/dotnet-cli.test.js
@@ -2,16 +2,24 @@
 // #region Imports
 // -----------------------------------------------------------------------------------------
 
-const child_process = require("child_process");
 const dotnetBuild = require("./dotnet-build");
 const dotnetCli = require("./dotnet-cli");
 const dotnetPath = require("./dotnet-path");
 const faker = require("faker");
-const path = require("path");
 const shell = require("shelljs");
 const testUtils = require("../tests/test-utils");
 
 // #endregion Imports
+
+// -----------------------------------------------------------------------------------------
+// #region Mocks
+// -----------------------------------------------------------------------------------------
+
+// Mocking dir module to suppress lots of extra output from popd/pushd errors from lack of actual
+// directory stack.
+jest.mock("./dir");
+
+// #endregion Mocks
 
 // -----------------------------------------------------------------------------------------
 // #region Tests

--- a/modules/dotnet-cli.test.js
+++ b/modules/dotnet-cli.test.js
@@ -23,12 +23,12 @@ describe("dotnetCli", () => {
     // -----------------------------------------------------------------------------------------
 
     describe("cmd", () => {
-        test("given a space separated string of arguments, it returns a properly command string to the Cli executable", () => {
+        test("given a string array of arguments, it returns a properly command string to the Cli executable", () => {
             // Arrange
             const cliPath = testUtils.randomFile();
-            const cliArgs = faker.random.words(3);
+            const cliArgs = faker.random.words(3).split(" ");
             jest.spyOn(dotnetPath, "cliPath").mockImplementation(() => cliPath);
-            const expectedString = `dotnet ${cliPath} ${cliArgs}`;
+            const expectedString = `dotnet ${cliPath} ${cliArgs.join(" ")}`;
 
             // Act
             const result = dotnetCli.cmd(cliArgs).toString();
@@ -50,10 +50,10 @@ describe("dotnetCli", () => {
             shellExitSpy = jest.spyOn(shell, "exit").mockImplementation();
         });
 
-        test("when dotnetPath.cliDir returns undefined, it calls dotnetBuild.run", () => {
+        test("when dotnetPath.cliPath returns undefined, it calls dotnetBuild.run", () => {
             // Arrange
             const dotnetPathSpy = jest
-                .spyOn(dotnetPath, "cliDir")
+                .spyOn(dotnetPath, "cliPath")
                 .mockImplementation(() => undefined);
             const dotnetBuildSpy = jest
                 .spyOn(dotnetBuild, "run")

--- a/modules/dotnet-path.js
+++ b/modules/dotnet-path.js
@@ -15,7 +15,7 @@ const upath = require("upath");
 // -----------------------------------------------------------------------------------------
 
 /**
- * Representse the directory where a Presentation.Cli dll would live after a `dotnet build`
+ * Represents the directory where a Presentation.Cli dll would live after a `dotnet build`
  */
 const cliFilePath = "Presentation/Cli/bin/Debug/**/*Cli.dll";
 

--- a/modules/dotnet.test.js
+++ b/modules/dotnet.test.js
@@ -12,6 +12,16 @@ const testUtils = require("../tests/test-utils");
 // #endregion Imports
 
 // -----------------------------------------------------------------------------------------
+// #region Mocks
+// -----------------------------------------------------------------------------------------
+
+// Mocking dir module to suppress lots of extra output from popd/pushd errors from lack of actual
+// directory stack.
+jest.mock("./dir");
+
+// #endregion Mocks
+
+// -----------------------------------------------------------------------------------------
 // #region Tests
 // -----------------------------------------------------------------------------------------
 

--- a/package.json
+++ b/package.json
@@ -14,6 +14,7 @@
     ],
     "scripts": {
         "configure:git": "echo Ensuring git hooksPath is set to .githooks directory; git config core.hooksPath .githooks; chmod +x .githooks/*",
+        "postpublish": "cross-env-shell \"git add -A && git commit -m \"$npm_package_version\" && git push origin master\"",
         "preinstall": "npm run configure:git",
         "test": "jest",
         "test:unit": "jest --coverage '^((?!and-cli).)*$'",

--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
         "cli"
     ],
     "scripts": {
-        "configure:git": "echo Ensuring git hooksPath is set to .githooks directory && git config core.hooksPath .githooks && chmod +x .githooks/*",
+        "configure:git": "echo Ensuring git hooksPath is set to .githooks directory; git config core.hooksPath .githooks; chmod +x .githooks/*",
         "preinstall": "npm run configure:git",
         "test": "jest",
         "test:unit": "jest --coverage '^((?!and-cli).)*$'",

--- a/package.json
+++ b/package.json
@@ -17,8 +17,8 @@
         "postpublish": "cross-env-shell \"git add -A && git commit -m \"$npm_package_version\" && git push origin master\"",
         "preinstall": "npm run configure:git",
         "test": "jest",
-        "test:unit": "jest --coverage '^((?!and-cli).)*$'",
-        "test:integration": "jest --runInBand '(and-cli)(.)*(.test.js)'",
+        "test:unit": "jest --coverage \"^((?!and-cli).)*$\"",
+        "test:integration": "jest --runInBand \"(and-cli)(.)*(.test.js)\"",
         "watch:test": "jest --watch"
     },
     "repository": {

--- a/tests/test-utils.js
+++ b/tests/test-utils.js
@@ -246,16 +246,6 @@ const testUtils = {
     },
 
     /**
-     * Wrapper of faker.random.number
-     *
-     * @param {number} [min=Number.MIN_VALUE] Minimum number value to randomly generate
-     * @param {number} [max=Number.MAX_VALUE] Maximum number value to randomly generate
-     */
-    randomNumber(min = Number.MIN_VALUE, max = Number.MAX_VALUE) {
-        return faker.random.number({ min, max });
-    },
-
-    /**
      * Implementation of a random file generating function, since faker.js does not seem to have
      * it yet
      *
@@ -272,6 +262,16 @@ const testUtils = {
         }
 
         return `${this.randomWord()}${extension}`;
+    },
+
+    /**
+     * Wrapper of faker.random.number
+     *
+     * @param {number} [min=Number.MIN_VALUE] Minimum number value to randomly generate
+     * @param {number} [max=Number.MAX_VALUE] Maximum number value to randomly generate
+     */
+    randomNumber(min = Number.MIN_VALUE, max = Number.MAX_VALUE) {
+        return faker.random.number({ min, max });
     },
 
     /**

--- a/tests/test-utils.js
+++ b/tests/test-utils.js
@@ -260,15 +260,18 @@ const testUtils = {
      * it yet
      *
      * See https://stackoverflow.com/a/39960706
-     * @param {string} extension
+     * @param {boolean} [includePath=false] Flag to determine whether a full path is generated along
+     * with the filename, or just the filename itself.
      * @returns
      */
-    randomFile(extension) {
-        if (StringUtils.isEmpty(extension)) {
-            extension = faker.system.commonFileExt();
+    randomFile(includePath = false) {
+        const extension = `.${faker.system.commonFileExt()}`;
+
+        if (includePath) {
+            return `${this.randomPath()}${extension}`;
         }
 
-        return `${this.randomPath()}.${extension}`;
+        return `${this.randomWord()}${extension}`;
     },
 
     /**

--- a/tests/test-utils.js
+++ b/tests/test-utils.js
@@ -246,6 +246,54 @@ const testUtils = {
     },
 
     /**
+     * Wrapper of faker.random.number
+     *
+     * @param {number} [min=Number.MIN_VALUE] Minimum number value to randomly generate
+     * @param {number} [max=Number.MAX_VALUE] Maximum number value to randomly generate
+     */
+    randomNumber(min = Number.MIN_VALUE, max = Number.MAX_VALUE) {
+        return faker.random.number({ min, max });
+    },
+
+    /**
+     * Implementation of a random file generating function, since faker.js does not seem to have
+     * it yet
+     *
+     * See https://stackoverflow.com/a/39960706
+     * @param {string} extension
+     * @returns
+     */
+    randomFile(extension) {
+        if (StringUtils.isEmpty(extension)) {
+            extension = faker.system.commonFileExt();
+        }
+
+        return `${this.randomPath()}.${extension}`;
+    },
+
+    /**
+     * Implementation of a random path generating function based on https://stackoverflow.com/a/39960706
+     *
+     * @param {number} [maxDepth=5] Maximum number of directories deep to generate a path for.
+     * @returns A randomly generated path with the platform-specific separator
+     */
+    randomPath(maxDepth = 5) {
+        // Override invalid values for the max depth of the path
+        if (maxDepth == null || maxDepth <= 1) {
+            maxDepth = 5;
+        }
+
+        const words = [];
+        for (let i = 0; i < maxDepth; i++) {
+            words.push(this.randomWord());
+        }
+        const randomPath = words.join(path.sep).toLowerCase();
+        return path.format({
+            base: randomPath,
+        });
+    },
+
+    /**
      * Wrapper of faker.random.word.
      *
      * Unfortunately there is an unresolved bug https://github.com/Marak/faker.js/issues/661

